### PR TITLE
add: handle hard links when copying with .dockerignore

### PR DESCRIPTION
**Pull request by nalind**
Detect files that are hard linked together, and use that information to avoid copying their contents more often than we have to.

### CODE REVIEW

**Comment by rhatdan** [comment]
@QiWang19 We might need similar fixes for podman cp.

File: util/util_unix.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+type hardlinkDeviceAndInode struct {
+	device, inode uint64
+}
+
+type HardlinkChecker struct {
+	hardlinks sync.Map
+}
+
+func (h *HardlinkChecker) Check(fi os.FileInfo) string {
+	if stat, ok := fi.Sys().(*syscall.Stat_t); ok && stat.Nlink > 1 {
```
Comment (rhatdan) [inline util/util_unix.go]: Should you avoid directories?

File: util.go [hunk 2]
```diff
@@ -144,22 +146,34 @@ func (b *Builder) copyFileWithTar(tarIDMappingOptions *IDMappingOptions, chownOp
...
+				hdr.Linkname = linkname
+			} else {
+				hardlinkChecker.Add(fi, chrootedDest)
 			}
-			defer func() {
-				if f != nil {
-					if err := f.Close(); err != nil {
-						logrus.Debugf("error closing %s: %v", f.Name(), err)
-					}
+			if hdr.Typeflag == tar.TypeReg {
```
Comment (rhatdan) [inline util.go]: Is this check necessary, isnce you have checked it above?

Comment (nalind) [reply util.go]: We might be changing it from TypeReg to TypeLink at line 159.

File: util/util_unix.go (hunk 1 above)
Comment (nalind) [reply util/util_unix.go]: For correctness, yes.  We shouldn't call this except for regular files, but still.

File: util.go (hunk 2 above)
Comment (rhatdan) [reply util.go]: Then why not put this in else block above.

Comment (nalind) [reply util.go]: Good point, changing it.

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**